src/components/Input/index.tsx

```diff
@@ -0,0 +1,184 @@
+import { Transition } from '@headlessui/react';
+import { InformationCircleIcon } from '@heroicons/react/outline';
+import {
+  ExclamationCircleIcon,
+  EyeIcon,
+  EyeOffIcon,
+} from '@heroicons/react/solid';
+import classNames from 'classnames';
+import React, { useEffect, useState, ChangeEventHandler } from 'react';
+
+export interface InputProps {
+  type: 'text' | 'number' | 'password';
+  onChange: ((v: string) => void) | ((v: number) => void);
+  value: string | number;
+  label?: string;
+  description?: string;
+  placeholder?: string;
+  required?: boolean;
+  maxLength?: number;
+  min?: number;
+  max?: number;
+  focusColor?: string;
+  error?: string;
+  integerOnly?: boolean;
+}
+
+export const Input: React.FC<InputProps> = ({
+  type = 'text',
+  onChange,
+  value,
+  label,
+  description,
+  placeholder,
+  required,
+  maxLength,
+  min = Number.MIN_SAFE_INTEGER, // If a value gets below this, it will lose precision, looking strange to the user
+  max = Number.MAX_SAFE_INTEGER, // If a value gets above this, it will lose precision, looking strange to the user
+  focusColor = 'focus:ring-blue-300',
+  error,
+  integerOnly,
+}) => {
+  const [tooltip, setTooltip] = useState('');
+  const [tooltipShow, setTooltipShow] = useState(false);
+  const [passwordVisibility, setPasswordVisibility] = useState(false);
+  let keyPressed: string;
+
+  useEffect(() => {
+    if (tooltipShow) {
+      const timeout = setTimeout(() => {
+        setTooltipShow(false);
+      }, 2500);
+      return () => clearTimeout(timeout);
+    }
+    return;
+  }, [tooltipShow]);
+
+  const handleChange: ChangeEventHandler<HTMLInputElement> = (e) => {
+    e.preventDefault();
+    let val = e.target.value;
+    if (type === 'number') {
+      if (keyPressed === '-') val = (-value).toString();
+      val = (+val).toString();
+      if (min && +val < min) {
+        setTooltip(`Your number must be ${min} or above`);
+        setTooltipShow(true);
+        return;
+      }
+      if (max && +val > max) {
+        setTooltip(`Your number must be ${max} or under`);
+        setTooltipShow(true);
+        return;
+      }
+
+      // Prevents input from going into scientific notation
+      if (+val < 1e-6 && +val > -1e-6 && val !== '0' && val !== '') {
+        setTooltip('Your number is far too precise');
+        setTooltipShow(true);
+        return;
+      }
+    } else {
+      if (maxLength && val.length > maxLength) {
+        setTooltip(`Your text can be at most ${maxLength} characters long`);
+        setTooltipShow(true);
+        return;
+      }
+    }
+    onChange((type === 'number' ? +val : val) as never);
+  };
+
+  return (
+    <div className="relative">
+      {label && (
```

Comment:
have you tested this to check if it renders a `0`? Sometimes when this label is undefined, the resulting DOM becomes a 0.
The correct way to handle this would be `label ? .... : null`

Reply:
Yeah I tested. It only does that if that variable was the number 0